### DISCUSSION TITLE
Top N Aggregates

### DISCUSSION THREAD

**Discussion by davidkohn88** [timescale/timescaledb-toolkit#40] [Feature Discussion]
Original issue in #38: 
In situations where one has zipfian or other long tail distributions of items, one often wants an aggregate that computes the topN most frequent values and lops off the rest for storage and lookup efficiency reasons. Take the case where you're monitoring traffic by URL and want to show the top 100 URLs by traffic in a day, there might be millions of URLs that are hit, but most of them only have a few hits and what you care about are the top 100. An aggregate that stored the top 100 urls and their counts on a daily basis and then you could re aggregate across multiple days if needed (though it would be an approximation of the true top100, it would be relatively accurate). I think this could be significantly more useful than the count min sketch in #6 .

**Comment by davidkohn88** [comment]
What would the best representation of this be? Would we do compression on the data in the aggregate?

**Reply by JLockerman** [reply]
One tempting option is to try to store fixed-size arrays of `(count, value)` ordered by decreasing count. This would allow finding the top-n elements very quickly, but merging is not trivial; the obvious may is to temporarily store a `value -> count` map, update all the counts, re-sort the tuples, and throw out the smallest values. We'll have to looks at the related works to see if there're better things to do.

**Reply by JLockerman** [reply]
http://archive.dimacs.rutgers.edu/Workshops/WGUnifyingTheory/Slides/cormode.pdf looks like a decent overview of various algorithms

**Comment by davidkohn88** [comment]
What would the API look like? I assume we would unnest the data? Would we only support text values (and just have people cast ints or whatever else to that?) and would you unnest the results? Could you downsample and just get the top10 from a top100 agg?

**Comment by JLockerman** [comment]
Some links that might be relevant:

- http://archive.dimacs.rutgers.edu/Workshops/WGUnifyingTheory/Slides/cormode.pdf
- http://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
- https://github.com/apache/druid/issues/7187
- https://github.com/ozturkosu/cms_topn
- https://zpjiang.me/2017/11/13/top-k-elementes-system-design/
- http://web.stanford.edu/class/cs369g/files/lectures/lec7.pdf
